keps/sig-api-machinery/4355-coordinated-leader-election/README.md

```diff
@@ -0,0 +1,1107 @@
+- Once one of the components updates its identity lease with new version
+  information, the coordinated election controller will preempt the current
+  leader so that this lower version component becomes leader.
+- When the remaining two nodes can rollback, the first node will typically
+  remain leader, but if a new election occurs, the available older version
+  components will be elected.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+#### Risk: Amount of writes performed by leader election increases substantially
+
+This enhancement introduces an identity lease for each instance of each component.
+
+Example:
+
+- HA cluster with 3 control plane nodes
+- 3 elected components (kube-controller-manager, schedule, cloud-controller-manager) per control plane node
+- 9 identity leases are created and renewed by the components
+
+Introducing this feature is roughtly equivalent to adding the same lease load
+as adding 9 nodes to a kubernetes cluster.
+
+The [API Server Identity enhancement](../1965-kube-apiserver-identity) also
+introduces similar leases. For comparison, in a HA cluster with 3 control plane
+nodes, API Server Identity adds 3 leases.
+
+This risk can be migitated by scale testing and, if needed, extending the lease
+duration and renewal times to reduce writes/s.
+
+#### Risk: leases watches increases apiserver load substantially
+
+The [Unknown Version Interoperability Proxy (UVIP) enhancement](../4020-unknown-version-interoperability-proxy) also adds
+lease watches on [API Server Identity](../1965-kube-apiserver-identity) leases in the kube-system namespace.
+This enhancement would increase the expected number of resources being watched from ~3 (for UVIP) to ~12.
+
+#### Risk: We have to "start over" and build confidence in a new leader election algorithm
+
+We've built confidence in the existing leasing algorithm, through an investment
+of engineering effort, and in core hours testing it and running it in production.
+
+Changing the algorithm "resets the clock" and forces us to rebuild confidence on
+the new algorithm.
+
+The goal of this proposal is to minimize this risk by reusing as much of the
+existing lease algorithm as possible:
+
+- Renew leases in exactly the same way as before
+- Leases can never be claimed by another leader until a lease expires
+
+#### Risk: How is the election controller elected?
+
+It's not. It will run directly in the apiserver. The apiserver runs very few controllers, and they are not elected, but instead all run concurrently in HA configurations.  
+Requires the election controller make careful use concurrency control primitives to ensure multiple instances collaborate, not fight.*
```

Comment:
i feel that today the fight happens only once based on optimistic concurrency. That allows more quick response for candidates. In a centrally coordinated system, the candidates may churn for some reason and the central component may lag significantly.

Reply:
This is a good point.  

Some algorithms add a "election term" which delays the election by a lease duration. If we did this we'd double our expected recovery time.

I'm hoping to avoid that and just have the election controller use it's current known list of candidates (based on lease watches) at the time of lease expiration to immediately selected a new leader. and claim the lease on that leaders behalf  There is then a short delay while the selected leader watches the lease and realizes it's been elected.

Given that the current election approach is polling.. if we instead use watches we might actually be faster (but not faster than if we switched the existing approach to use watches).